cpp/src/umap/knn_graph/algo.cuh

```diff
@@ -57,25 +91,56 @@ inline void launcher(const raft::handle_t& handle,
                      const ML::UMAPParams* params,
                      cudaStream_t stream)
 {
-  std::vector<float*> ptrs(1);
-  std::vector<int> sizes(1);
-  ptrs[0]  = inputsA.X;
-  sizes[0] = inputsA.n;
-
-  raft::spatial::knn::brute_force_knn(handle,
-                                      ptrs,
-                                      sizes,
-                                      inputsA.d,
-                                      inputsB.X,
-                                      inputsB.n,
-                                      out.knn_indices,
-                                      out.knn_dists,
-                                      n_neighbors,
-                                      true,
-                                      true,
-                                      static_cast<std::vector<int64_t>*>(nullptr),
-                                      params->metric,
-                                      params->p);
+  if (params->build_algo == ML::UMAPParams::graph_build_algo::BRUTE_FORCE_KNN) {
+    std::vector<float*> ptrs(1);
+    std::vector<int> sizes(1);
+    ptrs[0]  = inputsA.X;
+    sizes[0] = inputsA.n;
+
+    raft::spatial::knn::brute_force_knn(handle,
+                                        ptrs,
+                                        sizes,
+                                        inputsA.d,
+                                        inputsB.X,
+                                        inputsB.n,
+                                        out.knn_indices,
+                                        out.knn_dists,
+                                        n_neighbors,
+                                        true,
+                                        true,
+                                        static_cast<std::vector<int64_t>*>(nullptr),
+                                        params->metric,
+                                        params->p);
+  } else {  // nn_descent
+    RAFT_EXPECTS(static_cast<size_t>(n_neighbors) <= params->nn_descent_params.graph_degree,
+                 "n_neighbors should be smaller than the graph degree computed by nn descent");
+
+    auto graph = get_graph_nnd(handle, inputsA, params);
+
+    auto indices_d = raft::make_device_matrix<int64_t, int64_t>(
+      handle, inputsA.n, params->nn_descent_params.graph_degree);
+
+    raft::copy(indices_d.data_handle(),
+               graph.graph().data_handle(),
+               inputsA.n * params->nn_descent_params.graph_degree,
+               stream);
+
+    raft::matrix::slice_coordinates coords{static_cast<int64_t>(0),
+                                           static_cast<int64_t>(0),
+                                           static_cast<int64_t>(inputsA.n),
+                                           static_cast<int64_t>(n_neighbors)};
+
+    if (graph.distances().has_value()) {
```

Comment:
We don't always need distances in this use-case?

Reply:
We do, but because the distances is `optional` type, fails during runtime with bad optional access if we don't check for whether it has a value or not!

Reply:
Perhaps this should be a RAFT_EXPECTS instead of an if condition then? Just to safeguard that we throw an error if a user mistakenly uses nn_descent through python but sets return_distances=False